In [13]:
train.data <- read.csv(file.path("..", "data", "training_data.csv"))
test.data <- read.csv(file.path("..", "data", "test_data.csv"))

In [4]:
set.seed(100)
numeric.intensity <- as.numeric(train.data$Intensity)
x <- train.data[, -c(2,3)]
x$Intensity <- numeric.intensity
n.before <- dim(x)[2] #numbers of predictors before the reduction
idx.zero.var <- apply(x, 2, var) == 0
x <- x[,!idx.zero.var]
y <- train.data$VALENCE.PLEASANTNESS

## Boosting 1

In [5]:
#searching for input variables with zero variance (without the varible Intensity)
set.seed(100)
x <- train.data[, -c(1,2,3)]

idx.zero.var <- apply(x, 2, var) == 0
x <- x[,!idx.zero.var]


#attache Intensity as factor
x$Intensity <- as.factor(train.data$Intensity)

data <- x
data$VALENCE.PLEASANTNESS <- train.data$VALENCE.PLEASANTNESS


In [7]:
set.seed(100)
#train and validation indexes
len <- length(x[,1])
idx.train <- sample(1:len, 2*len/3)

#xgboost does not accept data frames therefore we will first convert the data into ordinary matrices
library(xgboost)
library(Matrix)
train.x <- sparse.model.matrix(VALENCE.PLEASANTNESS ~ . -1, data = data[idx.train,])
validation.x <- sparse.model.matrix(VALENCE.PLEASANTNESS ~ . -1, data = data[-idx.train,])
train.y <- data$VALENCE.PLEASANTNESS[idx.train]
validation.y <- data$VALENCE.PLEASANTNESS[-idx.train]

In [ ]:
boost.heart <- xgboost(train.x, label = train.y,
                      objective = "reg:squarederror",
                      eta = 0.01,
                      max_depth = 2,
                      nround = 500)

In [9]:
prediction.train <- predict(boost.heart, train.x)
prediction.validation <- predict(boost.heart, validation.x)
MSE.train <- mean((prediction.train - train.y)^2)
MSE.validation <- mean((prediction.validation - validation.y)^2)


In [10]:
MSE.train
MSE.validation

[1] 300.1458

[1] 439.2583

In [81]:
library(xgboost)
library(Matrix)
#Boosting Submission
set.seed(100)
#Preparation of training and test data
train <- train.data[, -c(1,2,3)]
idx.zero.var <- apply(train, 2, var) == 0

train <- train[,!idx.zero.var]
test <- test.data[,-c(1,2)]
test <- test[,!idx.zero.var]


#test$Intensity <- as.factor(test.data$Intensity)
train$Intensity <- as.numeric(train.data$Intensity)-1
train
#test intensity is always at level high, so that the prediction function has a problem (cheat with adding a row that afterwards is substracted)
#test <- rbind(test, train[1,])
train.x = train

train$VALENCE.PLEASANTNESS <- train.data$VALENCE.PLEASANTNESS
train.y = train$VALENCE.PLEASANTNESS

train.x <- sparse.model.matrix(VALENCE.PLEASANTNESS ~ . -1, data = train)
#test.x <- sparse.model.matrix(VALENCE.PLEASANTNESS ~ . -1, data = test)
train.y <- train$VALENCE.PLEASANTNESS
#validation.y <- data$VALENCE.PLEASANTNESS[-idx.train]

complexity.from.pubmed MW     AMW    Sv     Se     Sp     Si     Mv   
1   302.0                  208.33 5.952  19.698 34.491 21.523 39.571 0.563
2    72.6                  122.18 6.431  11.349 18.745 12.261 21.285 0.597
3   259.0                  242.29 7.572  20.832 32.167 21.693 35.535 0.651
4    56.6                   88.12 6.294   7.537 14.189  7.955 16.080 0.538
5    84.7                  136.21 6.191  12.876 21.629 14.023 24.701 0.585
6   160.0                  192.28 6.409  17.644 29.723 19.000 33.740 0.588
7   116.0                  170.22 7.401  15.349 22.745 16.261 25.285 0.667
8    25.0                  114.26 4.395  12.741 24.952 14.853 29.737 0.490
9   355.0                  268.30 7.666  21.304 36.025 21.614 39.950 0.609
10   68.4                  152.26 9.516   9.911 16.342 11.250 17.920 0.619
11  155.0                  148.17 7.798  12.537 19.189 12.955 21.080 0.660
12   25.1                   88.17 4.898   8.876 17.629 10.023 20.701 0.493
13   35.2                  102.20 4.867  10.402 20.513 11.784 24.116 0.495
14  101.0                  134.19 6.710  12.349 19.745 13.261 22.285 0.617
15  114.0                  146.16 7.308  11.493 20.727 11.625 22.914 0.575
16   52.5                  130.26 4.824  13.456 26.280 15.307 30.946 0.498
17  112.0                  160.24 5.935  14.359 27.051 15.455 30.950 0.532
18   68.9                  116.18 5.809  10.590 19.956 11.477 22.910 0.530
19   31.0                   60.06 7.508   4.483  8.422  4.432  9.249 0.560
20   55.4                  108.15 6.759   9.822 15.862 10.500 17.870 0.614
21   46.0                  112.24 4.677  12.214 23.069 14.091 27.322 0.509
22  101.0                  122.13 8.142  10.010 15.305 10.193 16.664 0.667
23   62.8                  108.15 6.759   9.822 15.862 10.500 17.870 0.614
24   83.0                  124.15 7.303  10.537 17.189 10.955 19.080 0.620
25  103.0                  138.18 6.909  12.064 20.073 12.716 22.495 0.603
26   19.6                   74.14 4.943   7.349 14.745  8.262 17.285 0.490
27  160.0                  192.28 6.409  17.644 29.723 19.000 33.740 0.588
28  270.0                  196.32 5.774  18.698 33.491 20.523 38.571 0.550
29  103.0                  115.15 6.774   9.558 17.291  9.960 19.578 0.562
30  153.0                  214.39 5.229  21.278 40.141 23.807 46.816 0.519
... ...                    ...    ...    ...    ...    ...    ...    ...  
679  89.0                  170.33  5.010 17.510 33.047 19.830 38.777 0.515
680  17.5                  122.28  7.643  9.008 15.571 11.102 17.916 0.563
681  59.1                  102.15  6.009  9.064 17.073  9.716 19.495 0.533
682 156.0                  178.25  6.602 16.117 26.840 17.239 30.325 0.597
683 173.0                  214.31  7.654 17.019 28.244 18.341 31.455 0.608
684  49.5                   88.12  6.294  7.537 14.189  7.955 16.080 0.538
685  76.4                  112.19  5.610 10.876 19.629 12.023 22.701 0.544
686 143.0                  170.28  5.676 16.171 29.607 17.762 34.156 0.539
687  32.9                   86.15  5.384  8.349 15.745  9.262 18.285 0.522
688  89.3                  144.24  5.548 13.644 25.723 15.000 29.740 0.525
689 128.0                  166.25  6.394 14.918 25.833 16.034 29.697 0.574
690 156.0                  178.25  6.602 16.117 26.840 17.239 30.325 0.597
691 117.0                  150.15  7.507 11.208 21.055 11.079 23.123 0.560
692 101.0                  120.16  7.068 10.822 16.862 11.500 18.870 0.637
693  12.4                  126.29 11.481  7.142 10.880  9.227 12.006 0.649
694 150.0                  154.28  5.320 15.456 28.280 17.307 32.946 0.533
695  10.3                   44.06  6.294  3.768  7.095  3.977  8.040 0.538
696  11.1                   90.21  6.014  7.821 14.494  9.455 16.996 0.521
697 177.0                  162.20  7.373 14.064 22.073 14.716 24.495 0.639
698  26.5                   80.10  8.010  6.570 10.087  6.773 11.412 0.657
699  43.2                   82.11  6.843  7.295 11.978  7.739 13.455 0.608
700 108.0                  15

In [82]:
test <- test.data[,-c(1,2)]
test <- test[,!idx.zero.var]
test$Intensity <- as.numeric(test.data$Intensity)-1

In [83]:
boost.heart <- xgboost(train.x, label = train.y,
                      objective = "reg:squarederror",
                      eta = 0.01,
                      max_depth = 2,
                      nround = 500)

[1]	train-rmse:48.543175 
[2]	train-rmse:48.156307 
[3]	train-rmse:47.774090 
[4]	train-rmse:47.396358 
[5]	train-rmse:47.023838 
[6]	train-rmse:46.655029 
[7]	train-rmse:46.291256 
[8]	train-rmse:45.931023 
[9]	train-rmse:45.575901 
[10]	train-rmse:45.224209 
[11]	train-rmse:44.877495 
[12]	train-rmse:44.534111 
[13]	train-rmse:44.195663 
[14]	train-rmse:43.860435 
[15]	train-rmse:43.530098 
[16]	train-rmse:43.202881 
[17]	train-rmse:42.880520 
[18]	train-rmse:42.561157 
[19]	train-rmse:42.246597 
[20]	train-rmse:41.934952 
[21]	train-rmse:41.627995 
[22]	train-rmse:41.323921 
[23]	train-rmse:41.024517 
[24]	train-rmse:40.727852 
[25]	train-rmse:40.436672 
[26]	train-rmse:40.147301 
[27]	train-rmse:39.863300 
[28]	train-rmse:39.581905 
[29]	train-rmse:39.305069 
[30]	train-rmse:39.029663 
[31]	train-rmse:38.759762 
[32]	train-rmse:38.490467 
[33]	train-rmse:38.226208 
[34]	train-rmse:37.966202 
[35]	train-rmse:37.706909 
[36]	train-rmse:37.450603 
[37]	train-rmse:37.199436 
[38]	train

In [79]:
prediction.boost = predict(boost.heart, as.matrix(test))
submission <- data.frame(Id = 1:68, VALENCE.PLEASANTNESS = prediction.boost)
write.csv(submission, file = "../Submissions/boosting2.csv", row.names = FALSE)


## Boosting 2 - Regularized gradient boosting 

One difference between boosting and random forests: in boosting, because the growth of a particular tree takes into account the other trees that have already been grown, smaller trees are typically sufficient (less splits and depth)

In [3]:
library(xgboost)

In [1]:
set.seed(100)
len <- length(x[,1])
idx.train <- sample(1:len, 3*len/4)

train.x <- x[idx.train,]
train.y <- y[idx.train]
test.x <- x[-idx.train,]
test.y <- y[-idx.train]


ERROR: Error in eval(expr, envir, enclos): objet 'x' introuvable


In [5]:
dtrain = xgb.DMatrix(data =  as.matrix(train.x), label = train.y )
dtest = xgb.DMatrix(data =  as.matrix(test.x), label = test.y)

In [6]:
watchlist = list(train=dtrain, test=dtest)

In [52]:
bst = xgb.train(data = dtrain, 
                max.depth = 8, 
                eta = 0.3, 
                nthread = 2, 
                nround = 1000, 
                watchlist = watchlist, 
                objective = "reg:squarederror", 
                early_stopping_rounds = 50,
                print_every_n = 500)

[1]	train-rmse:36.493153	test-rmse:37.327053 
Multiple eval metrics are present. Will use test_rmse for early stopping.
Will train until test_rmse hasn't improved in 50 rounds.

Stopping. Best iteration:
[7]	train-rmse:10.149774	test-rmse:24.547579



We get a train-rmse of 10.15 and a test-rmse of 24.55. But, the parameters were chosen randomly
Now, let's tune the algorithm with 3 parameters : 
1) The number of trees 

2) The shrinkage parameter lambda : Typical values are 0.01 or 0.001, and the right choice can depend on the problem. Very small λ can require using a very large value of B in order to achieve good performance.

3) The number of splits in each tree, which controls the complexity of the boosted ensemble (controlled with max.depth)

Let's run a slower learning model, by reducing the learning rate, and reducing the number of splits in each tree

In [53]:
bst_slow = xgb.train(data = dtrain, 
                        max.depth=5, 
                        eta = 0.01, 
                        nthread = 2, 
                        nround = 10000, 
                        watchlist = watchlist, 
                        objective = "reg:squarederror", 
                        early_stopping_rounds = 50,
                        print_every_n = 500)

[1]	train-rmse:48.982822	test-rmse:47.107918 
Multiple eval metrics are present. Will use test_rmse for early stopping.
Will train until test_rmse hasn't improved in 50 rounds.

Stopping. Best iteration:
[280]	train-rmse:12.274206	test-rmse:23.773617



We can see that we reduced the test-rmse by 6%. The problem here : What we have done here is fit to the training set and the test set at the same time (leading to model overfit). 

 We need to work with a validation set and only at the end evaluate the model performance against the test set.

In [54]:
len <- length(train.x[,1])
idx.valid <- sample(1:len, 2*len/3)

train.val.x <- train.x[idx.valid,]
train.val.y <- train.y[idx.valid]

valid.x <- train.x[-idx.valid,]
valid.y <- train.y[-idx.valid]


In [55]:
gb_train = xgb.DMatrix(data = as.matrix(train.val.x), label = train.val.y )
gb_valid = xgb.DMatrix(data = as.matrix(valid.x), label = valid.y )


In [56]:
watchlist = list(train = gb_train, valid = gb_valid)

In [57]:
bst_slow = xgb.train(data= gb_train, 
                        max.depth = 10, 
                        eta = 0.01, 
                        nthread = 2, 
                        nround = 10000, 
                        watchlist = watchlist, 
                        objective = "reg:squarederror", 
                        early_stopping_rounds = 50,
                        print_every_n = 500)

[1]	train-rmse:49.436710	valid-rmse:48.084187 
Multiple eval metrics are present. Will use valid_rmse for early stopping.
Will train until valid_rmse hasn't improved in 50 rounds.

Stopping. Best iteration:
[305]	train-rmse:5.971423	valid-rmse:24.474022



In [58]:
y_hat_valid = predict(bst_slow, dtest)
test_mse = mean(((y_hat_valid - test.y)^2))
test_mse
test_rmse = sqrt(test_mse)
test_rmse

[1] 671.6173

[1] 25.91558

This is higher then on the first run, but we can be confident that the improved score is not due to overfit thanks to our use of a validation set! A lower rmse isn't necessarily better if it comes at the cose of overfit, we now have more confidence in external predictions.



Let's find the best hyper-parameter combinations

In [59]:
max.depths = c(7, 9)
etas = c(0.01, 0.001)

best_params = 0
best_score = 0

count = 1

for( depth in max.depths ) {
    for(num in etas) {

        bst_grid = xgb.train(data = gb_train, 
                                max.depth = depth, 
                                eta=num, 
                                nthread = 2, 
                                nround = 10000, 
                                watchlist = watchlist, 
                                objective = "reg:squarederror", 
                                early_stopping_rounds = 50, 
                                verbose=0)

        if(count == 1){
            best_params = bst_grid$params
            best_score = bst_grid$best_score
            count = count + 1
            }
        else if( bst_grid$best_score < best_score){
            best_params = bst_grid$params
            best_score = bst_grid$best_score
        }
    }
}

best_params
best_score

$max_depth
[1] 7

$eta
[1] 0.001

$nthread
[1] 2

$objective
[1] "reg:squarederror"

$validate_parameters
[1] TRUE

[1] 23.79161

In [60]:
bst_tuned = xgb.train( data = gb_train, 
                        max.depth = 7, 
                        eta = 0.001, 
                        nthread = 2, 
                        nround = 10000, 
                        watchlist = watchlist, 
                        objective = "reg:squarederror", 
                        early_stopping_rounds = 50,
                        print_every_n = 500)

y_hat_xgb_grid = predict(bst_tuned, dtest)

test_mse = mean(((y_hat_xgb_grid - test.y)^2))
test_rmse = sqrt(test_mse)

[1]	train-rmse:49.839478	valid-rmse:48.421314 
Multiple eval metrics are present. Will use valid_rmse for early stopping.
Will train until valid_rmse hasn't improved in 50 rounds.

[501]	train-rmse:32.446823	valid-rmse:34.793262 
[1001]	train-rmse:21.792664	valid-rmse:27.858894 
[1501]	train-rmse:15.200837	valid-rmse:24.959738 
[2001]	train-rmse:11.097453	valid-rmse:23.966742 
Stopping. Best iteration:
[2400]	train-rmse:8.857841	valid-rmse:23.791611



In [61]:
test_rmse

[1] 25.70601

We should now maybe try to submit to Kaggle to see if we get better on the test set